src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -139,13 +139,25 @@ static P11Key convertKey(Token token, Key key, String algo,
         if (key instanceof SecretKey == false) {
             throw new InvalidKeyException("Key must be a SecretKey");
         }
+
+        final String keyAlgo = key.getAlgorithm();
+        if (key instanceof PBEKey && keyAlgo != null && algo != null) {
+            P11Util.KDFData kdfData = P11Util.kdfDataMap.get(keyAlgo);
+            if (kdfData != null && (keyAlgo.equals(algo) ||
+                    kdfData.keyAlgo.equals(algo))) {
+                return key instanceof P11Key ?
+                        (P11Key) key : // already derived
```

Comment:
As far as I've reviewed, different algorithms conversion (such as `PBEWithHmacSHA1AndAES_128` &rarr; `PBEWithHmacSHA224AndAES_128`) isn't supported in _SunJCE_ or _SunPKCS11_ SKFs. But yes, we should go through the code below in case the `P11PBEKey` we are converting is from a different PKCS#11 token.

We'll adjust this in subsequent changes.